### PR TITLE
Support for Multiple Instances of an Icon in an Image

### PR DESCRIPTION
**Pull request by marilynzhang**
Functionality that allows for running on multiple instance of an icon in an image:
- Small tweaks to read in the TfRecord with a list of bounding boxes per image instead of one bounding box
- Generalized evaluation function that supports both single and multi-instances, by counting the number of false positives, false negatives, true positives, and true negatives, and calculating accuracy, recall, and precision

Continuous Integration Tests:
- gpylint passes locally 
- Unit tests for the general evaluation function that covers edge cases like empty lists or variable-length lists, and "double penalizing" for a proposed box that doesn't pass the IOU threshold and leaves a gold box without any matched proposed box -- this case is considered as both a false positive and a false negative.

### CODE REVIEW

File: modules/util.py [hunk 1]
```diff
@@ -1,16 +1,194 @@
...
+
+  Returns:
+      Tuple(Number of false pos, false neg, true pos, true neg)
+  """
+  num_false_pos = 0
+  num_false_neg = 0
+  num_true_pos = 0
+  num_true_neg = 0
+  for proposed_box_list, gold_box_list in zip(proposed_boxes, gold_boxes):
+    matches = {}
```
Comment (a5malik) [inline modules/util.py:117]: lets have comment specifying the keys and values.

"map from golden boxes to the iou of their best matching proposed box"

and also what len(matches) is supposed to represent.

Comment (a5malik) [reply modules/util.py:117]: but I also suggest to not use the box as the key but its index.

File: modules/util.py [hunk 2]
```diff
@@ -1,16 +1,194 @@
...
+    for proposed_box in proposed_box_list:
+      # find the gold box that maximizes IOU
+      max_iou = 0
+      max_gold_box = None
+      for gold_box in gold_box_list:
```
Comment (a5malik) [inline modules/util.py:122]: I recommend using indices as ids instead of the box.

so you can use 

for gold_index, gold_box in enumerate(gold_box_list):

File: modules/util.py [hunk 3]
```diff
@@ -1,16 +1,194 @@
...
+          max_gold_box = gold_box
+      # if the gold box already has a match, increase false positives
+      # (condition holds even if there are no gold boxes; max_gold_box = None)
+      proposed_iou = max_iou
+      if max_gold_box in matches:
+        num_false_pos += 1
+        current_max_iou = matches[max_gold_box]
+        if current_max_iou > proposed_iou:
+          proposed_iou = matches[max_gold_box]
+      matches[max_gold_box] = proposed_iou
```
Comment (a5malik) [inline modules/util.py:135]: hmm. if max_gold_box is None, because the max_iou was 0, does that mean we have a key of None, and matches[None] = 0? is that intentional?

File: modules/util.py [hunk 4]
```diff
@@ -1,16 +1,194 @@
...
+      gold_boxes: a list of BoundingBox lists, one for each image,
+       as the ground truth.
+
+  Returns:
+      Tuple(Number of false pos, false neg, true pos, true neg)
+  """
+  num_false_pos = 0
+  num_false_neg = 0
+  num_true_pos = 0
+  num_true_neg = 0
```
Comment (a5malik) [inline modules/util.py:115]: Lets have a pseudocode algorithm here in comments. It was a little hard to parse the code through all the ifs and nested for loops.

File: modules/util.py [hunk 5]
```diff
@@ -1,16 +1,194 @@
...
+      for gold_box in gold_box_list:
+        iou = proposed_box.calculate_iou(gold_box)
+        if iou > max_iou:
+          max_iou = iou
+          max_gold_box = gold_box
+      # if the gold box already has a match, increase false positives
+      # (condition holds even if there are no gold boxes; max_gold_box = None)
+      proposed_iou = max_iou
+      if max_gold_box in matches:
+        num_false_pos += 1
```
Comment (a5malik) [inline modules/util.py:131]: its a little weird to increment num_false_pos here and also add to it in the end with the rest of the assignments below.

How about we just have another descriptive variable, increment that, and add that to the assignment expression below, so all the updates are in one place?

File: modules/util.py [hunk 6]
```diff
@@ -1,16 +1,194 @@
...
+        current_max_iou = matches[max_gold_box]
+        if current_max_iou > proposed_iou:
```
Comment (ewadkins) [inline modules/util.py:133]: Could replace current_max_iou with matches[max_gold_box] directly. With the other comment I added regarding the name of matches, I think this is acceptable.

File: modules/util.py (hunk 1 above)
Comment (ewadkins) [inline modules/util.py:117]: Maybe rename matches to matchedGoldenToIou. Unless it's obvious what the values of a map are (IOU is not obvious), it's good to specify.

Also +1 to not using the boxes themselves as keys.

File: modules/util.py [hunk 7]
```diff
@@ -1,16 +1,194 @@
...
+      max_gold_box = None
+      for gold_box in gold_box_list:
+        iou = proposed_box.calculate_iou(gold_box)
+        if iou > max_iou:
+          max_iou = iou
+          max_gold_box = gold_box
+      # if the gold box already has a match, increase false positives
+      # (condition holds even if there are no gold boxes; max_gold_box = None)
+      proposed_iou = max_iou
+      if max_gold_box in matches:
```
Comment (ewadkins) [inline modules/util.py:130]: Can you add a comment for each conditional, to make the logic more clear on first glance. This condition in particular can surprise the reader, since something being a match wouldn't be a false positive, but this is specifically boxes that have already been matched.

File: modules/util.py [hunk 8]
```diff
@@ -1,16 +1,194 @@
...
+      # find the gold box that maximizes IOU
+      max_iou = 0
+      max_gold_box = None
+      for gold_box in gold_box_list:
+        iou = proposed_box.calculate_iou(gold_box)
+        if iou > max_iou:
+          max_iou = iou
+          max_gold_box = gold_box
+      # if the gold box already has a match, increase false positives
+      # (condition holds even if there are no gold boxes; max_gold_box = None)
```
Comment (ewadkins) [inline modules/util.py:128]: I would guard against the "max_gold_box is None" case directly with a conditional, to not only make it more clear to the reader what happens in that case, but also because setting a value for None in matches seems very unexpected behavior, and even though the algorithm works fine, it's best not to rely on your implementation being able to handle the strange "side effect" of None being in matches. Especially given that you are later using the length of matches.

File: modules/util.py [hunk 9]
```diff
@@ -1,16 +1,194 @@
...
+      matches[max_gold_box] = proposed_iou
+    # evaluate matches with double penalty for mismatches
+    current_true_pos = np.sum(
+        [1 if iou >= iou_threshold else 0 for iou in matches.values()])
+    num_true_pos += current_true_pos
+    num_false_pos += len(matches) - current_true_pos
+    num_false_neg += len(matches) - current_true_pos
+    num_false_neg += len(gold_box_list) - len(matches)
+    if not proposed_box_list and not gold_box_list:
+      num_true_neg += 1
```
Comment (ewadkins) [inline modules/util.py:144]: This section would be ideal for commenting. So much easier to understand through comments. Also the None key in matches worries me given that you're using len(matches). Is there a one-off error here?

File: modules/util.py [hunk 10]
```diff
@@ -1,16 +1,194 @@
...
+      gold_box = BoundingBox(xmin, ymin, xmax, ymax)
+      single_image_gold_boxes.append(gold_box)
+    all_images_gold_boxes.append(single_image_gold_boxes)
+  return all_images_gold_boxes
+
+
+def parse_images_and_icons(
+    parsed_image_dataset: tf.python.data.ops.dataset_ops.MapDataset
+) -> Tuple[List[np.ndarray], List[np.ndarray]]:
+  """Private function for parsing images and icons.
```
Comment (ewadkins) [inline modules/util.py:76]: (nit) Not private anymore.

Also, maybe something more descriptive of the behavior, like "Returns the images and icons in the given dataset".

Alternatively, and this may actually be a better pattern to follow, you could have get_image_icon_pairs, which returns a list of (image, icon) tuples, which would clearly indicate that the images and icons are corresponding.

File: modules/util.py (hunk 3 above)
Comment (marilynzhang) [reply modules/util.py:135]: Removed this.

File: modules/bounding_box.py [hunk 11]
```diff
@@ -2,23 +2,23 @@
...
+  min_y: int
+  max_x: int
+  max_y: int
 
   def __init__(self, min_x: float, min_y: float, max_x: float, max_y: float):
-    self.min_x = min_x
-    self.min_y = min_y
-    self.max_x = max_x
-    self.max_y = max_y
+    object.__setattr__(self, "min_x", int(min_x))
```
[the remark this answers is not in the document]
Comment (marilynzhang) [reply modules/bounding_box.py:18]: This was because we were directly hashing the bounding boxes in the implementation of the evaluation metric (needed Bounding Box to be frozen for it to be hashable, which mean setting attribute in this non-pretty way) -- now that we're just using the index, the Bounding Box dataclass can be unfrozen and this will be reverted! :)

File: tests/unit_test.py [hunk 12]
```diff
@@ -1,28 +1,102 @@
...
+                      (_BOX_E, _BOX_F, 1 / 7), (_BOX_F, _BOX_G, 2 / 6),
+                      (_BOX_E, _BOX_G, 2 / 6)]
 
 
 @pytest.mark.parametrize("box_1,box_2,expected", bounding_box_tests)
 def test_iou(box_1, box_2, expected):
   assert box_1.calculate_iou(box_2) == expected
 
 
+box_list_1 = [[_BOX_A, _BOX_B], [_BOX_C, _BOX_D]]
```
Comment (a5malik) [inline tests/unit_test.py:37]: capitalize these?

File: modules/util.py [hunk 13]
```diff
@@ -1,16 +1,242 @@
...
+  else:
+    recall = num_true_pos / (num_true_pos + num_false_neg)
+
+  if output_path:
+    with open(output_path, "a") as output_file:
+      output_file.write("Accuracy: %f\n" % accuracy)
+      output_file.write("Precision: %f\n" % precision)
+      output_file.write("Recall: %f\n" % recall)
+
+  print("Accuracy: %f\n" % accuracy)
```
Comment (a5malik) [inline modules/util.py:235]: maybe in a future PR, you could change all these to logging.debug() or logging.info() wherever appropriate. See python's logging package:

https://docs.python.org/3/howto/logging.html

Comment (marilynzhang) [reply modules/util.py:235]: This does seem helpful, thank you! Noted.
